vignettes/R-dev-blog-post.qmd

```diff
@@ -0,0 +1,100 @@
+
+```{r, eval = FALSE}
+library(reticulate)
+
+# Adding a NumPy array and an R array dispatches to the default
+# `+` method for "python.builtin.object"
+np_array(1:3) + array(1:3)
+
+# Adding a TensorFlow Tensor and an R array invokes
+# invokes the specific `+` method for "tensorflow.tensor"
+array(1:3) + tensorflow::as_tensor(1:3)
+
+# Adding a NumPy array with a TensorFlow Tensor.
+# Prior to R 4.3.0, this would signal an error, Warning about
+#   Incompatible methods ("+.tensorflow.tensor", "+.python.builtin.object")
+# Beginning with R 4.3.0, chooseOpsMethod() is called to choose the appropriate
+# Ops method, and this now works
+np_array(1:3) + as_tensor(1:3)
+```
+
+### `nameOfClass()`
+
+Though S7 uses S3 dispatch, the class name used for S3 dispatch is considered an implementation detail (it is non-syntatic and mangled to minimize the potential for conflicts, and not intended to be user-facing). Additionally, it can be subject to change if the generic definition migrates between packages.
+
+For these reasons, it is desirable to be able to check for inheritance without directly using the S3 class name. To enable this, beginning with R 4.3.0 `base::inherits()` now accepts arbitrary objects for the second argument. If the second argument is an S3 object, the new `nameOfClass()` S3 generic is invoked to resolve the appropriate class name.
+
+This permits usage in S7 like this:
+
+```{r, eval = FALSE}
+ClassX <- S7::new_class("ClassX")
+
+x <- ClassX()
+
+inherits(x, ClassX)
+```
+
+This also has utility for other R packages, in particular, packages where an external class hierarchy is being mirrored in R. For example, in reticulate, the S3 class vector for Python object is generated from the Python module where the class is defined. However, in sophisticated codebases like Keras and TensorFlow, the location where a class is defined in the Python sources is considered an internal implementation detail, and often changes between library versions (i.e., the module where a class is defined is decoupled from the module where users are meant to access the class). To ensure compatibility between and across Python library versions, reticulate now implements a `nameOfClass()` method which enables R usage like this:
+
+```{r, eval = FALSE}
+inherits(x, tensorflow::tf$Variable)
+```
+
+### `@`
+
+Prior to R 4.3.0, `@` could be used only with S4 objects. Beginning with R 4.3.0, this operator can now perform dispatch S3 dispatch as well. This bring `@` to parity with `@<-` (which has long been able to do S3 dispatch), and makes `@` a peer (in dispatch-ability) of `$`. This is a strictly backwards compatible change, as S3 dispatch is only performed *after* S4 routines have had an opportunity to dispatch.
```

Comment:
Maybe note that there's still no default method for `@` but it's available for use in packages?